### ISSUE TITLE
[issue] KubeEdge didn't support Weave CNI 

### ISSUE THREAD

**Issue by FengGaoCSC** [kubeedge/kubeedge#3935]
**What happened**:
KubeEdge didn't support Weave CNI plugin 

**What you expected to happen**:

Except KubeEdge support Weave CNI as KubeEdge can support flannel CNI with issue #2677

**How to reproduce it (as minimally and precisely as possible)**:
`my environment :`

```
root@k8s-131:/home/k8s/kubeedge-1.9.2/keadm-v1.9.2-linux-amd64/keadm# kubectl get node -o wide
NAME         STATUS   ROLES                  AGE     VERSION                    INTERNAL-IP     EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION      CONTAINER-RUNTIME
master-131   Ready    control-plane,master   3h44m   v1.23.4                    192.168.0.131   <none>        Ubuntu 20.04 LTS     5.13.0-28-generic   docker://20.10.7
raspi1       Ready    agent,edge             17m     v1.22.6-kubeedge-v1.10.0   172.100.1.2     <none>        Ubuntu 20.04.3 LTS   5.4.0-1058-raspi    docker://20.10.7
raspi2       Ready    agent,edge             18m     v1.22.6-kubeedge-v1.10.0   172.100.0.2     <none>        Ubuntu 20.04.1 LTS   5.4.0-1065-raspi    docker://20.10.7
```
`After deployment the weavenet cni`
```
root@k8s-131:/home/k8s/kubeedge-1.9.2/keadm-v1.9.2-linux-amd64/keadm# kubectl get pod --all-namespaces -o wide
NAMESPACE     NAME                                 READY   STATUS    RESTARTS      AGE     IP              NODE         NOMINATED NODE   READINESS GATES
kube-system   coredns-64897985d-5h6zk              1/1     Running   0             57m     10.32.0.1       master-131   <none>           <none>
kube-system   coredns-64897985d-dvnj4              1/1     Running   0             57m     10.32.0.2       master-131   <none>           <none>
kube-system   etcd-master-131                      1/1     Running   1             3h47m   192.168.0.131   master-131   <none>           <none>
kube-system   kube-apiserver-master-131            1/1     Running   1             3h47m   192.168.0.131   master-131   <none>           <none>
kube-system   kube-controller-manager-master-131   1/1     Running   1             3h47m   192.168.0.131   master-131   <none>           <none>
kube-system   kube-proxy-czpx2                     1/1     Running   0             21m     172.100.1.2     raspi1       <none>           <none>
kube-system   kube-proxy-ml7n9                     1/1     Running   0             3h47m   192.168.0.131   master-131   <none>           <none>
kube-system   kube-proxy-n8sr8                     1/1     Running   0             21m     172.100.0.2     raspi2       <none>           <none>
kube-system   kube-scheduler-master-131            1/1     Running   1             3h47m   192.168.0.131   master-131   <none>           <none>
kube-system   weave-net-hf82x                      2/2     Running   0             86s     192.168.0.131   master-131   <none>           <none>
kube-system   weave-net-l2jcn                      0/2     Error     6 (33s ago)   86s     172.100.0.2     raspi2       <none>           <none>
kube-system   weave-net-xtr5v                      0/2     Error     6 (27s ago)   86s     172.100.1.2     raspi1       <none>           <none>
```

`and after check the description of weave pod we can see that`

```sh
// status of weave pod on raspi board
root@k8s-131:/home/k8s/keadm-v1.9.2-linux-amd64/keadm# kubectl get pod --all-namespaces -o wide
NAMESPACE     NAME                             READY   STATUS    RESTARTS   AGE   IP              NODE     NOMINATED NODE   READINESS GATES
.................
kube-system   weave-net-2m4b9                  0/2     "Error"     360        17h   172.100.1.2     raspi1   <none>           <none>
kube-system   weave-net-kfwbt                  2/2     Running   0          17h   192.168.0.131   master   <none>           <none>
.................

//root@k8s-131:/home/k8s/keadm-v1.9.2-linux-amd64/keadm# kubectl describe pod weave-net-2m4b9 -n kube-system
Name:                 weave-net-2m4b9
Namespace:            kube-system
Priority:             2000001000
Priority Class Name:  system-node-critical
Node:                 raspi1/172.100.1.2
..................
Containers:
  weave:
    Container ID:  docker://284c3cf268c09b624583391e79050eaf90564dcda194d022adc30f963df5ff80
    Image:         ghcr.io/weaveworks/launcher/weave-kube:2.8.1
    Image ID:      docker-pullable://ghcr.io/weaveworks/launcher/weave-kube@sha256:d797338e7beb17222e10757b71400d8471bdbd9be13b5da38ce2ebf597fb4e63
    Port:          <none>
    Host Port:     <none>
    Command:
      /home/weave/launch.sh                 "<------ here launch.sh failed to get KUBE_PEERS name"
    State:          Terminated
      Reason:       Error
      Exit Code:    1                                  "<------- exit code = 1"
      Started:      Wed, 15 Jun 2022 09:28:55 +0800
................
```
` and only weave's pause container is launched`

```sh
root@raspi-work-1:/home/ubuntu/keadm-v1.9.2-linux-arm64/keadm# docker ps
CONTAINER ID   IMAGE                COMMAND                  CREATED        STATUS        PORTS     NAMES
68c219987500   a7521b8c1d65         /usr/local/bin/kube…   17 hours ago   Up 17 hours             k8s_kube-proxy_kube-proxy-8bjjs_kube-system_14890e53-cc85-46e1-82ec-1951849d55e0_0
50ed4c2c23b3   kubeedge/pause:3.1   /pause                 17 hours ago   Up 17 hours             k8s_POD_weave-net-2m4b9_kube-system_e30de19c-1685-429e-9d2c-6ed5b761681f_0
0fd5f1a65ade   kubeedge/pause:3.1   /pause                 17 hours ago   Up 17 hours             k8s_POD_kube-proxy-8bjjs_kube-system_14890e53-cc85-46e1-82ec-1951849d55e0_0
```
`by check the code of weave https://github.com/weaveworks/weave/blob/master/prog/weave-kube/launch.sh`

```sh
......
if [ -z "$KUBE_PEERS" ]; then
    if ! KUBE_PEERS=$(/home/weave/kube-utils); then
        echo Failed to get peers >&2
        exit 1                                         "<-------- here is the exit point that weave container failed to launch"
    fi
fi

peer_count() {
    echo $#
}
.........
post_start_actions &

/home/weave/weaver $EXTRA_ARGS --port=6783 $(router_bridge_opts) \
     --name="$PEERNAME" \
     --http-addr=$HTTP_ADDR $STATUS_OPTS --docker-api='' --no-dns \
     --db-prefix="$DB_PREFIX" \
     --ipalloc-range=$IPALLOC_RANGE $NICKNAME_ARG \
     --ipalloc-init $IPALLOC_INIT \
     --conn-limit=$CONN_LIMIT \
     $WEAVE_NPC_OPTS \
     $NO_MASQ_LOCAL_OPT \
     "$@" \
     $KUBE_PEERS
```
`and the weave's kube-util binary need to access api-server to get the cluster's all node list. source are on https://github.com/weaveworks/weave/blob/master/prog/kube-utils/main.go`

```go
// return the IP addresses of all nodes in the cluster
func getKubePeers(c kubernetes.Interface, includeWithNoIPAddr bool) ([]nodeInfo, error) {
	ctx := context.Background()
	nodeList, err := c.CoreV1().Nodes().List(ctx, api.ListOptions{})         <--- needs to access api-server. 
	if err != nil {
		return nil, err
	}
	addresses := make([]nodeInfo, 0, len(nodeList.Items))
	for _, peer := range nodeList.Items {
		var internalIP, externalIP string
		for _, addr := range peer.Status.Addresses {
			// Check it's a valid ipv4 address
			ip := net.ParseIP(addr.Address)
			if ip == nil || ip.To4() == nil {
				continue
			}
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
  - 1.21.4
  -
- KubeEdge version(e.g. `cloudcore --version` and `edgecore --version`):
  - 1.9.2
  - 
- <details><summary>Cloud nodes Environment:</summary>

  - Hardware configuration (e.g. `lscpu`):
  - OS (e.g. `cat /etc/os-release`):
     - ubuntu20.04
  - Kernel (e.g. `uname -a`):
  - Go version (e.g. `go version`):
  - Others:

  </details>
- <details><summary>Edge nodes Environment:</summary>

  - edgecore version (e.g. `edgecore --version`):
  - Hardware configuration (e.g. `lscpu`):
    - raspi
  - OS (e.g. `cat /etc/os-release`):
     - - ubuntu20.04
  - Kernel (e.g. `uname -a`):
  - Go version (e.g. `go version`):
  - Others:

  </details>

**Comment by fujitatomoya** [comment]
CC: @fisherxu

**Comment by vincentgoat** [comment]
Hi @fujitatomoya, as you describe above, the weave pod can't access api-server cause this error. Can you set the api-server address to the local meta server address (default is 127.0.0.1:10550). Also, the prerequisite is to turn on the meta server module in the edgecore.yaml and dynamicController module in the cloudcore.yaml.

**Comment by FengGaoCSC** [comment]
hi @vincentgoat


> Hi @fujitatomoya, as you describe above, the weave pod can't access api-server cause this error. Can you set the api-server address to the local meta server address (default is 127.0.0.1:10550). Also, the prerequisite is to turn on the meta server module in the edgecore.yaml and dynamicController module in the cloudcore.yaml.

-  I see the issue#2677 that flannel can set the the local meta server address (default is 127.0.0.1:10550) with following 
```yaml
- args:
        - --ip-masq
        - --kube-subnet-mgr
        - --kube-api-url=http://127.0.0.1:10550          <---- here you change the default local meta server address 
        name: kube-flannel-edge
 ```
but the `--kube-api-url=http://127.0.0.1:10550`  arguments is one of options of  `flanneld`.  and `Weave CNI` plugin haven't this options, Could you tell me how to set the api-server address to the local meta server address(default 127.0.0.1:10550)?

very thanks

**Comment by vincentgoat** [comment]
Which is the api-server that the `Weave CNI` use? Can you check it in the weave log files?

**Comment by fujitatomoya** [comment]
@vincentgoat are you suggesting that it weave init should issue the API to get `KUBE_PEERS` to the local meta server address (default is 127.0.0.1:10550)? instead of k8s api-server.

@FengGaoCSC i am not sure how this is doable though. any thoughts? probably we need to adjust the weave implementation or init script or something?

> I assume that weave uses the IP of the default kubernetes service.

this seems to be correct.

**Comment by FengGaoCSC** [comment]
@vincentgoat @fujitatomoya

hi vincentgoat:

I back to this issue, since weave CNI applied to kubedge edge node with the [weave's code](https://github.com/weaveworks/weave/blob/6d7a29ade6e196e51a800b9df658c8a017f1b38b/vendor/k8s.io/client-go/rest/config.go)  was changed to following to force the weave & weave NPC to access http://127.0.0.1:10550, the weave and weave NPC with running state but weave CNI still not work . 

Could you take a look at this issue how to enable weave cni work with kubeedge?

```diff
diff --git a/Makefile b/Makefile
index c903bf5f..fff1354c 100644
--- a/Makefile
+++ b/Makefile
@@ -3,7 +3,7 @@
 
 # If you can use docker without being root, you can do "make SUDO="
 SUDO=$(shell docker info >/dev/null 2>&1 || echo "sudo -E")
-BUILD_IN_CONTAINER=true
+BUILD_IN_CONTAINER=false
 RM=--rm
 RUN_FLAGS=-ti
 COVERAGE=
@@ -142,7 +142,8 @@ TEST_TLS_EXE=test/tls/tls
 NETWORKTESTER_EXE=test/images/network-tester/webserver
 
 # All binaries together in a list
-EXES=$(WEAVER_EXE) $(SIGPROXY_EXE) $(KUBEUTILS_EXE) $(WEAVENPC_EXE) $(WEAVEWAIT_EXE) $(WEAVEWAIT_NOOP_EXE) $(WEAVEWAIT_NOMCAST_EXE) $(WEAVEUTIL_EXE) $(RUNNER_EXE) $(TEST_TLS_EXE) $(NETWORKTESTER_EXE)
+#EXES=$(WEAVER_EXE) $(SIGPROXY_EXE) $(KUBEUTILS_EXE) $(WEAVENPC_EXE) $(WEAVEWAIT_EXE) $(WEAVEWAIT_NOOP_EXE) $(WEAVEWAIT_NOMCAST_EXE) $(WEAVEUTIL_EXE) $(RUNNER_EXE) $(TEST_TLS_EXE) $(NETWORKTESTER_EXE)
+EXES= $(KUBEUTILS_EXE) $(WEAVENPC_EXE) 
 
 # These stamp files are used to mark the current state of the build; whether an image has been built or not
 BUILD_UPTODATE=.build.uptodate
diff --git a/prog/kube-utils/main.go b/prog/kube-utils/main.go
index 42414626..5b244e1d 100644
--- a/prog/kube-utils/main.go
+++ b/prog/kube-utils/main.go
@@ -23,6 +23,7 @@ import (
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
 
 	weaveapi "github.com/weaveworks/weave/api"
 	"github.com/weaveworks/weave/common"
@@ -59,9 +60,9 @@ func getKubePeers(c kubernetes.Interface, includeWithNoIPAddr bool) ([]nodeInfo,
 		// Fallback for cases where a Node has an ExternalIP but no InternalIP
 		if internalIP != "" {
 			// exclude self from the list of peers this node will peer with
-			if isLocalNodeIP(internalIP) {
-				continue
-			}
+		//	if isLocalNodeIP(internalIP) {
+		//		continue
+		//	}
 			addresses = append(addresses, nodeInfo{name: peer.Name, addr: internalIP})
 		} else if externalIP != "" {
 			addresses = append(addresses, nodeInfo{name: peer.Name, addr: externalIP})
@@ -332,12 +333,24 @@ func main() {
 	flag.StringVar(&logLevel, "log-level", "info", "logging level (debug, info, warning, error)")
 	flag.Parse()
 
+	var cfg *rest.Config
+	var err error
+
 	common.SetLogLevel(logLevel)
-	config, err := rest.InClusterConfig()
+
+	//modified here to use the kubeedge AKE feature with url http://127.0.0.1:10550
+	/*config, err := rest.InClusterConfig()
 	if err != nil {
 		common.Log.Fatalf("[kube-peers] Could not get cluster config: %v", err)
+	}*/
+
+	cfg, err = clientcmd.BuildConfigFromFlags("http://127.0.0.1:10550", "" )
+
+	if err != nil {
+		common.Log.Fatalf("[kube-peers] KubeEdge weave Could not create kubeconfig: %v", err)
 	}
-	c, err := kubernetes.NewForConfig(config)
+
+	c, err := kubernetes.NewForConfig(cfg)
 	if err != nil {
 		common.Log.Fatalf("[kube-peers] Could not make Kubernetes connection: %v", err)
 	}
diff --git a/prog/weave-npc/main.go b/prog/weave-npc/main.go
index cfc48145..4c0ffedb 100644
--- a/prog/weave-npc/main.go
+++ b/prog/weave-npc/main.go
@@ -18,6 +18,7 @@ import (
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/cache"
+        "k8s.io/client-go/tools/clientcmd"
 
 	"github.com/weaveworks/weave/common"
 	"github.com/weaveworks/weave/common/chains"
@@ -246,10 +247,14 @@ func root(cmd *cobra.Command, args []string) {
 		common.Log.Fatalf("Failed to start ulogd: %v", err)
 	}
 
-	config, err := rest.InClusterConfig()
+	var cfg *rest.Config
+        var err error
+
+	//config, err := rest.InClusterConfig()
+	cfg, err = clientcmd.BuildConfigFromFlags("http://127.0.0.1:10550", "" )
 	handleFatal(err)
 
-	client, err := kubernetes.NewForConfig(config)
+	client, err := kubernetes.NewForConfig(cfg)
 	handleFatal(err)
 
 	ipt, err := iptables.New()
```

**Comment by vincentgoat** [comment]
Hi @FengGaoCSC, sure, I will give it a try.

**Comment by FengGaoCSC** [comment]
> Hi @FengGaoCSC, sure, I will give it a try.

thanks & very appreciate :)

**Comment by fujitatomoya** [comment]
@FengGaoCSC what about creating PR based on your suggestion, so that we can leave the comments.

**Comment by FengGaoCSC** [comment]
@fujitatomoya 

my suggestion of code change is based on `weaver CNI` source code but not `KubeEdge`, and this change still not work yet

**Comment by fujitatomoya** [comment]
@FengGaoCSC 

having the KubeEdge specific patch would be burden for KubeEdge community, so it would be nice to have common way to address our issue? just idea, i am not even sure if this is doable or not.

**Comment by FengGaoCSC** [comment]
> @FengGaoCSC
> 
> having the KubeEdge specific patch would be burden for KubeEdge community, so it would be nice to have common way to address our issue? just idea, i am not even sure if this is doable or not.

ok. got it and I will take care of this .

**Comment by vincentgoat** [comment]
Here is my processing:
1. Cherry-pick PR https://github.com/kubeedge/kubeedge/pull/3583 to workaround error `[kube-peers] Could not get Kubernetes version:`, because KubeEdge doesn't support `version` API currently. But there is one bug in this PR, note that to remove `|| !info.IsResourceRequest` in function `Generate`:
![image](https://user-images.githubusercontent.com/49225405/188420381-efd8a8f1-2f99-4dd5-9367-03d3080859ee.png)
I have already contacted the author of this PR to correct it.

2. Then I run `weave` successfully but the connection is blocked when I try to curl the service cluster IP. The details in the log of container `weave-npc` are as follows:
`{"log":"WARN: 2022/09/05 09:39:01.492699 TCP connection from 10.32.0.3:40028 to 10.32.0.4:80 blocked by Weave NPC.\n","stream":"stderr","time":"2022-09-05T09:39:01.492843591Z"}`

3. Some troubleshooting in the official DOC:
https://www.weave.works/docs/net/latest/kubernetes/kube-addon/#blocked-connections
Sadly, I tried this but still does not work.

I will continue to work out this issue. Please let me know if you make any progress.

**Comment by vincentgoat** [comment]
cc @Poorunga

**Comment by FengGaoCSC** [comment]
> Here is my processing:
> I will continue to work out this issue. Please let me know if you make any progress.

@vincentgoat 

thanks very much. I also working on this, if any progress, we can share it

**Comment by vincentgoat** [comment]
It works after set `disable-npc` to `true`, here is how I install weave: `kubectl apply -f "https://cloud.weave.works/k8s/net?k8s-version=$(kubectl version | base64 | tr -d '\n')&disable-npc=true"`

If an unauthorized error occurred, please cherry-pick this PR https://github.com/kubeedge/kubeedge/pull/3807 to fix it.

@FengGaoCSC

**Comment by FengGaoCSC** [comment]
@vincentgoat 

very appreciate. I will try your ways, any question I will keep contact with you.

**Comment by FengGaoCSC** [comment]
> It works after set `disable-npc` to `true`, here is how I install weave: `kubectl apply -f "https://cloud.weave.works/k8s/net?k8s-version=$(kubectl version | base64 | tr -d '\n')&disable-npc=true"`
> 
> If an unauthorized error occurred, please cherry-pick this PR #3807 to fix it.
> 
> @FengGaoCSC

I saw that PR https://github.com/kubeedge/kubeedge/pull/3807 
it solved the wrong service account in pod so we didn't need to change or force the weave to curl  http://127.0.0.1:10550. 

I will have try it.

thanks very much

**Comment by FengGaoCSC** [comment]
@vincentgoat  I try your method and the weave pod on edge's status was `pending`, maybe my test is not right. I will try it again.

**Comment by Poorunga** [comment]
#3583  needs to further optimize the code

**Comment by fujitatomoya** [comment]
off topic from this issue. but WeaveNet gone unmaintained, see https://github.com/weaveworks/weave/issues/3960.

and we have been working on community support via https://github.com/weaveworks/weave/issues/3989.

**Comment by fujitatomoya** [comment]
for [flannel work-around](https://github.com/kubeedge/kubeedge/issues/2677) (and https://github.com/kubeedge/kubeedge/issues/3935#issuecomment-1234962037) 
, it introduces the burden for orchestrator user.
which generates the new label `flannel-edge` to separate the deployment for `DaemonSets`.
i understand that this is work-around to deploy the flannel to be aware if the pod is deployed on cloud (direct api-server access) or edge(local meta server address 127.0.0.1:10550), although this should be concealed from user perspective.

that said, if this requires user's responsibility, every time user changes CNI, there would be specified work-around for each of CNI.

in architecture, CNI implementation should be agnostic from cloud or edge.
probably there would be common interface (environmental variable?) to pass the argument to let the application know where to access the 'api-server' for their aspect?
so that application can be agnostic from knowing 'api-server' location.

**Comment by fujitatomoya** [comment]
@FengGaoCSC @Windrow14 I will go ahead to close this issue since we ended up not using weavenet.

**Comment by fujitatomoya** [comment]
@FengGaoCSC can you close this? if any objections, i am okay to keep this open.

**Comment by FengGaoCSC** [comment]
@fujitatomoya 

yes we could close it

**Comment by FengGaoCSC** [comment]
Closed
